optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/decider/acceptor/greatdeluge/GreatDelugeAcceptor.java

```diff
@@ -0,0 +1,87 @@
+
+import org.optaplanner.core.api.score.Score;
+import org.optaplanner.core.impl.localsearch.decider.acceptor.AbstractAcceptor;
+import org.optaplanner.core.impl.localsearch.scope.LocalSearchMoveScope;
+import org.optaplanner.core.impl.localsearch.scope.LocalSearchPhaseScope;
+import org.optaplanner.core.impl.localsearch.scope.LocalSearchStepScope;
+import org.optaplanner.core.impl.score.ScoreUtils;
+
+public class GreatDelugeAcceptor extends AbstractAcceptor {
+
+    private Score initialWaterLevels = null;
+    private Score currentWaterLevel = null;
+
+    private Double rainSpeedRatio = DEFAULT_RAIN_SPEED_RATIO;
+    private Score rainSpeedScore = null;
+
+    // Good value to come out from. Source: https://github.com/UniTime/cpsolver from Tomas Muller
+    private static final double DEFAULT_RAIN_SPEED_RATIO = 0.99_999_995;
+
+
+    public Score getRainSpeedScore() {
+        return this.rainSpeedScore;
+    }
+
+    public void setRainSpeedScore(Score rainSpeedScore) {
+        this.rainSpeedScore = rainSpeedScore;
+        this.rainSpeedRatio = null;
+    }
+
+    public Score getInitialWaterLevels() {
+        return this.initialWaterLevels;
+    }
+
+    public void setInitialWaterLevels(Score initialLevel) {
+        this.initialWaterLevels = initialLevel;
+    }
+
+    public Double getRainSpeedRatio() {
+        return this.rainSpeedRatio;
+    }
+
+    public void setRainSpeedRatio(Double rainSpeedRatio) {
+        this.rainSpeedRatio = rainSpeedRatio;
+    }
+
+    public void phaseStarted(LocalSearchPhaseScope phaseScope) {
+        super.phaseStarted(phaseScope);
+
+        if (initialWaterLevels != null) {
+            for (double initialLevelLevel : ScoreUtils.extractLevelDoubles(initialWaterLevels)) {
+                if (initialLevelLevel < 0.0) {
+                    throw new IllegalArgumentException("The initial level (" + initialWaterLevels
+                                                               + ") cannot have negative level (" + initialLevelLevel + ").");
+                }
+            }
+            currentWaterLevel = initialWaterLevels;
+
+        } else {
+            currentWaterLevel = phaseScope.getBestScore().negate();
+        }
+    }
+
+    public void phaseEnded(LocalSearchPhaseScope phaseScope) {
+        super.phaseEnded(phaseScope);
+        initialWaterLevels = null;
+        rainSpeedRatio = DEFAULT_RAIN_SPEED_RATIO;
+        rainSpeedScore = null;
+    }
+
+    @Override
+    public boolean isAccepted(LocalSearchMoveScope moveScope) {
+
+        Score moveScore = moveScope.getScore();
+
+        return moveScore.compareTo(currentWaterLevel.negate()) >= 0;
+    }
+
+    public void stepStarted(LocalSearchStepScope stepScope) {
+        super.stepEnded(stepScope);
+        if (rainSpeedScore != null) {
+            currentWaterLevel = currentWaterLevel.subtract(rainSpeedScore);
+        } else {
+            currentWaterLevel = currentWaterLevel.multiply(rainSpeedRatio);
```

Comment:
Note (no changes needed): this inherently presumes a optimal score of Score.ZERO. 
In a future PR, after this one is merged, we'll want to relax this presumption, by adding a targetScore parameter.

- In most cases the perfect score [1] is ZERO, because most cases only have negative constraints. For example, in course scheduling, we'd love to see 0 hard constraints and 0 soft constraints broken. There are cases that mix in positive constraints (for example conference scheduling) and the optimal score could be 0hard/700soft. Possible fix: adding a targetScore parameter that defaults to Score.ZERO.

- In most cases, the optimal score is not the perfect score. For example, in VRP, a score 0hard/0soft is impossible, because the driving time factors into the soft score: delivering at 10 locations is impossible in 0 seconds. Possible fix: adding a targetScore parameter that defaults to Score.ZERO.

The code would look like this:

```
// Notice that the rainSpeedRatio's meaning changes: Instead of setting it to 0.95, we 'll set it to 0.05. [2]
Score increment = targetScore.subtract(currentWaterLevel).multiply(rainSpeedRatio);
currentWaterLevel += increment;
```

We don't know the optimal score in advance, so the targetScore could be a pita parameter if the default of ZERO isn't any good.

[1] perfect score is not general academic terminology. It's NOT a synonym for optimal score or lower/higher bound score.
[2] @triceo we need to figure this out before merging. I would expect that the ratio is the relative amount of the water that is added, between current water level and highest possible water level.

Reply:
On point we need to define clearly before merging What is the ratio? Is it 0.05 for 0.95? I would argue 0.05, as we increase the water level be 5% towards the final state of great deluge at every step. The idea of we keep 95% not flooded is inverse thinking. So I would argue that 95% is implementation friendly, and 5% is user friendly. We should use 5%.

Reply:
TODO for next pr: better name targetScore